launcher/agent/agent.go

```diff
@@ -152,10 +237,44 @@ func (a *agent) Attest(ctx context.Context, opts AttestAgentOpts) ([]byte, error
 	return resp.ClaimsToken, nil
 }
 
-func (a *agent) attest(nonce []byte, cel []byte) (*pb.Attestation, error) {
-	a.tpmMu.Lock()
-	defer a.tpmMu.Unlock()
-	return a.fetchedAK.Attest(client.AttestOpts{Nonce: nonce, CanonicalEventLog: cel, CertChainFetcher: http.DefaultClient})
+type tpmExtendAttester struct {
+	tpmMu     sync.Mutex
+	fetchedAK *client.Key
+	tpm       io.ReadWriteCloser
+}
+
+func (t *tpmExtendAttester) Extend(c cel.Content, l *cel.CEL) error {
+	return l.AppendEventPCR(t.tpm, cel.CosEventPCR, defaultCELHashAlgo, c)
+}
+
+func (t *tpmExtendAttester) Attest(nonce []byte) (interface{}, error) {
+	t.tpmMu.Lock()
+	defer t.tpmMu.Unlock()
+
+	return t.fetchedAK.Attest(client.AttestOpts{
+		Nonce:            nonce,
+		CertChainFetcher: http.DefaultClient,
+	})
+}
+
+type tdxExtendAttester struct {
+	tdxMu     sync.Mutex
+	qp        *tg.LinuxConfigFsQuoteProvider
+	tsmClient configfsi.Client
+}
+
+func (t *tdxExtendAttester) Extend(c cel.Content, l *cel.CEL) error {
+	return l.AppendEventRTMR(t.tsmClient, cel.CosRTMR, c)
+}
+
+func (t *tdxExtendAttester) Attest(nonce []byte) (interface{}, error) {
+	t.tdxMu.Lock()
+	defer t.tdxMu.Unlock()
+
+	var tdxNonce [tlabi.TdReportDataSize]byte
+	copy(tdxNonce[:], nonce)
+
+	return tg.GetQuote(t.qp, tdxNonce)
```

Comment:
nit: I'm more curious about whether this is a thread-safe operation.

Reply:
I assume the quote provider is thread safe (the underlying tsm will handle it via file lock maybe??) I'm not sure though. I added a `tdxMu` which should handle race condition from our context

Reply:
Good catch. Adding a `tdxMu` should handle the race condition.

tsm interface is not thread safe. https://lore.kernel.org/linux-coco/65397556ee5cc_7258329495@dwillia2-xfh.jf.intel.com.notmuch/

go-configfs has implemented the check (https://github.com/google/go-configfs-tsm/blob/main/report/report.go#L186). If multiple threads/processes writes to the same tsm entry, GetQuote will returns an error.